torch/nn/modules/module.py

```diff
@@ -1982,24 +1982,39 @@ def __delattr__(self, name):
             super().__delattr__(name)
 
     def _register_state_dict_hook(self, hook):
-        r"""Register a state-dict hook.
+        r"""Register a post-hook for the :meth:`~torch.nn.Module.state_dict` method.
 
-        These hooks will be called with arguments: `self`, `state_dict`,
-        `prefix`, `local_metadata`, after the `state_dict` of `self` is set.
-        Note that only parameters and buffers of `self` or its children are
-        guaranteed to exist in `state_dict`. The hooks may modify `state_dict`
-        inplace or return a new one.
+        It should have the following signature::
+            hook(module, state_dict, prefix, local_metadata) -> None or state_dict
+
+        The registered hooks can modify the ``state_dict`` inplace or return a new one.
+        If a new ``state_dict`` is returned, it will only be respected if it is the root
+        module that :meth:`~nn.Module.state_dict` is called from.
         """
         handle = RemovableHandle(self._state_dict_hooks)
         self._state_dict_hooks[handle.id] = hook
         return handle
 
+    def register_state_dict_post_hook(self, hook):
+        r"""Register a post-hook for the :meth:`~torch.nn.Module.state_dict` method.
+
+        It should have the following signature::
+            hook(module, state_dict, prefix, local_metadata) -> None
+
+        The registered hooks can modify the ``state_dict`` inplace.
+        """
+        handle = RemovableHandle(self._state_dict_hooks)
+        hook._from_public_api = True
```

Comment:
Can we add a bit of error checking here to ensure that even if the same function is used multiple times, it works ok.
I think the simplest thing to do here would be to say that if a function is used by the private API, we tag it, error if someone tries to use it with the public API and never clear that flag.
I think it's ok to say that if you use this weird private API there, will be restrictions.

This is a bit different from what you have here in the sense of:
- You want a _from_private_api flag
- Don't delete the flag on removal of the handle
- Assert the flag is not present in the public API
- Add the flag (even if already there) in the private API

wdyt?

Reply:
Makes sense, I did not think about the case of the same hook being registered multiple times

Reply:
Per offline discussion, add this attribute for the public method since overrides of this private method would not have the flag, and hence would not have the old behavior